Test failure: nodeos_read_terminate_at_block_if_lr_test

https://github.com/AntelopeIO/spring/actions/runs/14412295392/job/40423204461?pr=1364

```
Traceback (most recent call last):
  File "/__w/spring/spring/build/tests/nodeos_read_terminate_at_block_test.py", line 331, in <module>
    success = executeTest(
  File "/__w/spring/spring/build/tests/nodeos_read_terminate_at_block_test.py", line 69, in executeTest
    checkStatus(testNode, testNodeArgs)
  File "/__w/spring/spring/build/tests/nodeos_read_terminate_at_block_test.py", line 121, in checkStatus
    assert head == termAtBlock, f"head {head} termAtBlock {termAtBlock}"
AssertionError: head 101 termAtBlock 100
```

--- a/libraries/chain/controller.cpp
+++ b/libraries/chain/controller.cpp
@@ -2150,7 +2150,7 @@ struct controller_impl {
                if (auto head = fork_db.head(); head->block_num() > conf.truncate_at_block) {
                   ilog("Removing blocks past truncate-at-block ${t} from fork database with head at ${h}",
                         ("t", conf.truncate_at_block)("h", head->block_num()));
-                  fork_db.remove(conf.truncate_at_block);
+                  fork_db.remove(conf.truncate_at_block + 1);
                }
             });
          }

--- a/libraries/chain/controller.cpp
+++ b/libraries/chain/controller.cpp
@@ -2147,7 +2147,7 @@ struct controller_impl {
       if (conf.truncate_at_block > 0 && chain_head.is_valid()) {
          if (chain_head.block_num() == conf.truncate_at_block && fork_db_has_root()) {
             fork_db_.apply<void>([&](auto& fork_db) {
-               if (auto head = fork_db.head(); head->block_num() > conf.truncate_at_block) {
+               if (auto head = fork_db.head(); head && head->block_num() > conf.truncate_at_block) {
                   ilog("Removing blocks past truncate-at-block ${t} from fork database with head at ${h}",
                         ("t", conf.truncate_at_block)("h", head->block_num()));
                   fork_db.remove(conf.truncate_at_block + 1);

--- a/libraries/chain/controller.cpp
+++ b/libraries/chain/controller.cpp
@@ -2080,6 +2080,10 @@ struct controller_impl {
          db.undo();
       }
 
+      EOS_ASSERT(conf.terminate_at_block == 0 || conf.terminate_at_block > chain_head.block_num(),
+                 plugin_config_exception, "--terminate-at-block ${t} not greater than chain head ${h}",
+                 ("t", conf.terminate_at_block)("h", chain_head.block_num()));
+
       protocol_features.init( db );
 
       // At startup, no transaction specific logging is possible
@@ -2139,6 +2143,19 @@ struct controller_impl {
 
    ~controller_impl() {
       pending.reset();
+
+      if (conf.truncate_at_block > 0 && chain_head.is_valid()) {
+         if (chain_head.block_num() == conf.truncate_at_block && fork_db_has_root()) {
+            fork_db_.apply<void>([&](auto& fork_db) {
+               if (auto head = fork_db.head(); head->block_num() > conf.truncate_at_block) {
+                  ilog("Removing blocks past truncate-at-block ${t} from fork database with head at ${h}",
+                        ("t", conf.truncate_at_block)("h", head->block_num()));
+                  fork_db.remove(conf.truncate_at_block);
+               }
+            });
+         }
+      }
+
       //only log this not just if configured to, but also if initialization made it to the point we'd log the startup too
       if(okay_to_print_integrity_hash_on_stop && conf.integrity_hash_on_stop)
          ilog( "chain database stopped with hash: ${hash}", ("hash", calculate_integrity_hash()) );

--- a/libraries/chain/fork_database.cpp
+++ b/libraries/chain/fork_database.cpp
@@ -103,6 +103,7 @@ namespace eosio::chain {
       void             reset_root_impl( const bsp_t& root_bs );
       void             advance_root_impl( const block_id_type& id );
       void             remove_impl( const block_id_type& id );
+      void             remove_impl( block_num_type block_num );
       bsp_t            head_impl(include_root_t include_root) const;
       bool             set_pending_savanna_lib_id_impl(const block_id_type& id);
       branch_t         fetch_branch_impl( const block_id_type& h, uint32_t trim_after_block_num ) const;
@@ -586,6 +587,24 @@ namespace eosio::chain {
 
       for( const auto& block_id : remove_queue ) {
          index.erase( block_id );
+      }
+   }
+
+   template<class BSP>
+   void fork_database_t<BSP>::remove( block_num_type block_num ) {
+      std::lock_guard g( my->mtx );
+      return my->remove_impl( block_num );
+   }
+
+   template<class BSP>
+   void fork_database_impl<BSP>::remove_impl( block_num_type block_num ) {
+      // doesn't matter which index as there is no index over block_num
+      for (auto itr = index.template get<0>().begin(); itr != index.template get<0>().end(); ) {
+         if ((*itr)->block_num() >= block_num) {
+            itr = index.erase(itr);
+         } else {
+            ++itr;
+         }
       }
    }
 

--- a/libraries/chain/include/eosio/chain/controller.hpp
+++ b/libraries/chain/include/eosio/chain/controller.hpp
@@ -150,6 +150,7 @@ namespace eosio::chain {
             uint32_t                 maximum_variable_signature_length = chain::config::default_max_variable_signature_length;
             bool                     disable_all_subjective_mitigations = false; //< for developer & testing purposes, can be configured using `disable-all-subjective-mitigations` when `EOSIO_DEVELOPER` build option is provided
             uint32_t                 terminate_at_block     = 0;
+            uint32_t                 truncate_at_block      = 0;
             uint32_t                 num_configured_p2p_peers = 0;
             bool                     integrity_hash_on_start= false;
             bool                     integrity_hash_on_stop = false;

--- a/libraries/chain/include/eosio/chain/fork_database.hpp
+++ b/libraries/chain/include/eosio/chain/fork_database.hpp
@@ -82,6 +82,11 @@ namespace eosio::chain {
 
       void remove( const block_id_type& id );
 
+      /**
+       * Remove all blocks >= block_num
+       */
+      void remove( block_num_type block_num);
+
       bool is_valid() const; // sanity checks on this fork_db
 
       bool   has_root() const;

--- a/plugins/chain_plugin/chain_plugin.cpp
+++ b/plugins/chain_plugin/chain_plugin.cpp
@@ -410,9 +410,12 @@ void chain_plugin::set_program_options(options_description& cli, options_descrip
          ("delete-all-blocks", bpo::bool_switch()->default_value(false),
           "clear chain state database and block log")
          ("truncate-at-block", bpo::value<uint32_t>()->default_value(0),
-          "stop hard replay / block log recovery at this block number (if set to non-zero number)")
+          "Stop hard replay / block log recovery at this block number (if non-zero). "
+          "Can also be used with terminate-at-block to prune any received blocks from fork database on exit.")
          ("terminate-at-block", bpo::value<uint32_t>()->default_value(0),
-          "terminate after reaching this block number (if set to a non-zero number)")
+          "Stops the node after reaching the specified block number (if non-zero). "
+          "Use RPC endpoint /v1/producer/pause_at_block to pause at a specific block instead. "
+          "Combine with truncate-at-block to prune blocks beyond the specified number from the fork database on exit.")
          ("snapshot", bpo::value<std::filesystem::path>(), "File to read Snapshot State from")
          ;
 
@@ -483,9 +486,9 @@ namespace {
 
 void
 chain_plugin_impl::do_hard_replay(const variables_map& options) {
-         ilog( "Hard replay requested: deleting state database" );
-         clear_directory_contents( chain_config->state_dir );
-         auto backup_dir = block_log::repair_log( blocks_dir, options.at( "truncate-at-block" ).as<uint32_t>(), config::reversible_blocks_dir_name);
+   ilog( "Hard replay requested: deleting state database" );
+   clear_directory_contents( chain_config->state_dir );
+   auto backup_dir = block_log::repair_log( blocks_dir, options.at( "truncate-at-block" ).as<uint32_t>(), config::reversible_blocks_dir_name);
 }
 
 void chain_plugin_impl::plugin_initialize(const variables_map& options) {
@@ -686,8 +689,8 @@ void chain_plugin_impl::plugin_initialize(const variables_map& options) {
 
       chain_config->maximum_variable_signature_length = options.at( "maximum-variable-signature-length" ).as<uint32_t>();
 
-      if( options.count( "terminate-at-block" ))
-         chain_config->terminate_at_block = options.at( "terminate-at-block" ).as<uint32_t>();
+      chain_config->terminate_at_block = options.at( "terminate-at-block" ).as<uint32_t>();
+      chain_config->truncate_at_block = options.at( "truncate-at-block" ).as<uint32_t>();
 
       chain_config->num_configured_p2p_peers = options.count( "p2p-peer-address" );
 
@@ -766,26 +769,30 @@ void chain_plugin_impl::plugin_initialize(const variables_map& options) {
          EOS_THROW( extract_genesis_state_exception, "extracted genesis state from blocks.log" );
       }
 
+      uint32_t truncate_at_block = options.at( "truncate-at-block" ).as<uint32_t>();
+      uint32_t terminate_at_block = options.at( "terminate-at-block" ).as<uint32_t>();
+      if (truncate_at_block > 0 && terminate_at_block > 0) {
+         EOS_ASSERT(truncate_at_block == terminate_at_block, plugin_config_exception,
+                    "truncate-at-block ${a} must match terminate-at-block ${b}",
+                    ("a", truncate_at_block)("b", terminate_at_block));
+      } else if (truncate_at_block > 0 && !options.at( "hard-replay-blockchain" ).as<bool>()) {
+         wlog("truncate-at-block only applicable to --hard-replay-blockchain unless specified with --terminate-at-block");
+      }
+
       if( options.at( "delete-all-blocks" ).as<bool>()) {
          ilog( "Deleting state database and blocks" );
-         if( options.at( "truncate-at-block" ).as<uint32_t>() > 0 )
-            wlog( "The --truncate-at-block option does not make sense when deleting all blocks." );
          clear_directory_contents( chain_config->state_dir );
          clear_directory_contents( blocks_dir );
       } else if( options.at( "hard-replay-blockchain" ).as<bool>()) {
          do_hard_replay(options);
       } else if( options.at( "replay-blockchain" ).as<bool>()) {
          ilog( "Replay requested: deleting state database" );
-         if( options.at( "truncate-at-block" ).as<uint32_t>() > 0 )
-            wlog( "The --truncate-at-block option does not work for a regular replay of the blockchain." );
          if (!options.count( "snapshot" )) {
             auto first_block = block_log::extract_first_block_num(blocks_dir, retained_dir);
             EOS_ASSERT(first_block == 1, plugin_config_exception,
                        "replay-blockchain without snapshot requested without a full block log, first block: ${n}", ("n", first_block));
          }
          clear_directory_contents( chain_config->state_dir );
-      } else if( options.at( "truncate-at-block" ).as<uint32_t>() > 0 ) {
-         wlog( "The --truncate-at-block option can only be used with --hard-replay-blockchain." );
       }
 
       std::optional<chain_id_type> chain_id;

--- a/tests/nodeos_read_terminate_at_block_test.py
+++ b/tests/nodeos_read_terminate_at_block_test.py
@@ -60,7 +60,7 @@ def executeTest(cluster, testNodeId, testNodeArgs, resultMsgs):
         testNode = cluster.getNode(testNodeId)
         assert not testNode.verifyAlive() # resets pid so reluanch works
         peers = testNode.rmFromCmd('--p2p-peer-address')
-        testNode.relaunch(addSwapFlags={"--terminate-at-block": "9999999"})
+        testNode.relaunch(addSwapFlags={"--terminate-at-block": "0", "--truncate-at-block": "0"})
 
         # Wait for node to start up.
         time.sleep(3)
@@ -134,7 +134,7 @@ def checkReplay(testNode, testNodeArgs):
     ]))
 
     assert not testNode.verifyAlive()
-    testNode.relaunch(chainArg="--replay-blockchain", addSwapFlags={"--terminate-at-block": "9999999"})
+    testNode.relaunch(chainArg="--replay-blockchain", addSwapFlags={"--terminate-at-block": "0", "--truncate-at-block": "0"})
 
     # Wait for node to finish up.
     time.sleep(3)
@@ -213,7 +213,7 @@ def executeSnapshotBlocklogTest(cluster, testNodeId, resultMsgs, nodeArgs, termA
         assert testNode.kill(signal.SIGTERM)
 
     # Start from snapshot, replay through block log and terminate at specified block
-    chainArg=f'--snapshot {testNode.getLatestSnapshot()} --replay-blockchain --terminate-at-block {termAtBlock}'
+    chainArg=f'--snapshot {testNode.getLatestSnapshot()} --replay-blockchain --terminate-at-block {termAtBlock} --truncate-at-block {termAtBlock}'
     testNode.relaunch(chainArg=chainArg, waitForTerm=True)
 
     # Check the node stops at the correct block by checking the log.
@@ -268,10 +268,10 @@ try:
         0 : "--enable-stale-production"
     }
     regularNodeosArgs = {
-        1 : "--read-mode irreversible --terminate-at-block 100",
-        2 : "--read-mode head --terminate-at-block 125",
-        3 : "--read-mode speculative --terminate-at-block 150",
-        4 : "--read-mode irreversible --terminate-at-block 180"
+        1 : "--read-mode irreversible --terminate-at-block 100 --truncate-at-block 100",
+        2 : "--read-mode head --terminate-at-block 125 --truncate-at-block 125",
+        3 : "--read-mode speculative --terminate-at-block 150 --truncate-at-block 150",
+        4 : "--read-mode irreversible --terminate-at-block 180 --truncate-at-block 180"
     }
     replayNodeosArgs = {
         5 : "--read-mode irreversible",

--- a/unittests/fork_db_tests.cpp
+++ b/unittests/fork_db_tests.cpp
@@ -167,6 +167,19 @@ BOOST_FIXTURE_TEST_CASE(add_remove_test, generate_fork_db_state) try {
    BOOST_TEST(branch[1] == bsp11b);
 } FC_LOG_AND_RETHROW();
 
+BOOST_FIXTURE_TEST_CASE(remove_block_num_test, generate_fork_db_state) try {
+   BOOST_TEST(fork_db.size() == 14);
+   fork_db.remove(13); // remove all >= 13
+   BOOST_TEST(fork_db.size() == 8);
+
+   for (auto& i : all) {
+      if (i->block_num() < 13) {
+         BOOST_TEST(fork_db.get_block(i->id()) == i);
+      } else {
+         BOOST_TEST(!fork_db.get_block(i->id()));
+      }
+   }
+} FC_LOG_AND_RETHROW();
 
 // test `fork_database_t::validated_block_exists() const` member
 // -------------------------------------------------------------

--- a/unittests/fork_db_tests.cpp
+++ b/unittests/fork_db_tests.cpp
@@ -170,7 +170,7 @@ BOOST_FIXTURE_TEST_CASE(add_remove_test, generate_fork_db_state) try {
 BOOST_FIXTURE_TEST_CASE(remove_block_num_test, generate_fork_db_state) try {
    BOOST_TEST(fork_db.size() == 14);
    fork_db.remove(13); // remove all >= 13
-   BOOST_TEST(fork_db.size() == 8);
+   BOOST_TEST(fork_db.size() == 8); // 6 blocks >= 13
 
    for (auto& i : all) {
       if (i->block_num() < 13) {

